kore/src/Kore/Step/Axiom/Matcher.hs

```diff
@@ -178,7 +178,10 @@ matchIncremental termLike1 termLike2 =
                 $ MultiAnd.toPredicate predicate
             substitution' =
                 Predicate.fromSubstitution
-                $ Substitution.fromMap substitution
+                -- unsafeWrap is safe because the matcher guarantees the
+                -- substitution is normalized.
+                $ Substitution.unsafeWrap
```

Comment:
I actually liked that we had only one call to `unsafeWrap` in our non-testing code. Having the new checks in `unsafeWrap` certainly helps, but I'm going to make an argument that we still shouldn't use it. If, after reading it, you still think it's better this way, you can keep it as-is.

It does not seem obvious to me that "the matcher guarantees the substitution is normalized", and that it will always be so. Is there a strong argument for that?

I mean, maybe now it correctly matches `f(x, x)` to `f(g(a), g(b))` to get `x=g(a) and g(a)=g(b)`, and maybe it correctly matches `f(x, x)` and `f(y, C(y))`, where `C` is a constructor, to get `bottom`, and so on. However, it's not obvious that it does that (maybe it should be obvious, but I'm just missing something).

Even if it is obvious that it's already normalized, we may probably break this assumption at some point (intentionally or not), so why not go through the standard normalization? If it's about performance, we may consider adding a quick check in `normalizeSubstitution` to see whether everything is already normalized before doing anything else.

Reply:
> It does not seem obvious to me that "the matcher guarantees the substitution is normalized", and that it will always be so. Is there a strong argument for that?

Yes. [`substitute`](https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/Axiom/Matcher.hs#L445) performs an occurs-check (so there are no cyclic dependencies) and performs back-substitution (so no substitution depends on any other). For the sake of generating _matching_ solutions, it's also important that back-substitution is performed immediately: there is no possibility of generating duplicated substitutions (multiple substitutions for the same variable). We can't use the existing normalization for substitutions because it uses unification to de-duplicate substitutions.

If we ever changed this about the matcher, then we _must_ change this line too, and I would want it to throw an error otherwise.

Reply:
Ok, I see, thanks.